crates/proxy-server/src/mod_proxy.rs

```diff
@@ -0,0 +1,221 @@
+        };
+
+        // Log the listener address - this format is depended upon by integration tests
+        if self.use_tls {
+            tracing::info!("proxy listener (TLS) on {addr:?}");
+        } else {
+            tracing::info!("proxy listener on {addr:?}");
+        }
+
+        let params = Arc::new(self);
+
+        spawn(format!("proxy listener {addr:?}"), async move {
+            Self::accept_loop(listener, params, tls_acceptor).await
+        })?;
+
+        Ok(())
+    }
+
+    async fn accept_loop(
+        listener: TcpListener,
+        params: Arc<Self>,
+        tls_acceptor: Option<TlsAcceptor>,
+    ) {
+        // Get the local address from the listener
+        let local_address = match listener.local_addr() {
+            Ok(addr) => addr,
+            Err(err) => {
+                tracing::error!("failed to get local address: {err:#}");
+                return;
+            }
+        };
+
+        loop {
+            let (socket, peer_address) = match listener.accept().await {
+                Ok(tuple) => tuple,
+                Err(err) => {
+                    tracing::error!("accept failed: {err:#}");
+                    return;
+                }
+            };
+
+            let params = params.clone();
+            let tls_acceptor = tls_acceptor.clone();
+
+            tokio::spawn(async move {
+                let result = if let Some(acceptor) = tls_acceptor {
+                    match acceptor.accept(socket).await {
+                        Ok(tls_stream) => {
+                            Self::handle_client(tls_stream, peer_address, local_address, &params)
+                                .await
+                        }
+                        Err(err) => {
+                            tracing::debug!("TLS handshake failed from {peer_address:?}: {err:#}");
+                            return;
+                        }
+                    }
+                } else {
+                    Self::handle_client(socket, peer_address, local_address, &params).await
+                };
+
+                if let Err(err) = result {
+                    tracing::error!("proxy session error from {peer_address:?}: {err:#}");
+                }
+            });
+        }
+    }
+
+    async fn handle_client<S>(
+        stream: S,
+        peer_address: SocketAddr,
+        local_address: SocketAddr,
+        params: &ProxyListenerParams,
+    ) -> anyhow::Result<()>
+    where
+        S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    {
+        crate::proxy_handler::handle_proxy_client(
+            stream,
+            peer_address,
+            local_address,
+            params.timeout,
+            params.no_splice,
+            params.require_auth,
+        )
+        .await
+    }
+}
+
+/// Connection metadata passed to the proxy_server_auth_1929 callback
+#[derive(Clone, Debug, serde::Serialize)]
+pub struct ConnMeta {
+    pub peer_address: String,
+    pub local_address: String,
```

Comment:
you can have serde take care of serializing these to strings

```suggestion
    pub peer_address: SockAddr,
    pub local_address: SockAddr,
```